src/box/lua/schema.lua

```diff
@@ -3500,42 +3563,63 @@ local function revoke(uid, name, privilege, object_type, object_name, options)
     end
     local privilege_hex = privilege_check(privilege, object_type)
     options = options or {}
+    if options._origin == nil then
+        options._origin = DEFAULT_ORIGIN
+    elseif type(options._origin) ~= 'string' then
+        box.error(box.error.ILLEGAL_PARAMS, "options parameter '_origin' " ..
+                  "should be of type 'string'")
+    end
     local oid = object_resolve(object_type, object_name)
     local _priv = box.space[box.schema.PRIV_ID]
     local _vpriv = box.space[box.schema.VPRIV_ID]
     local tuple = _vpriv:get{uid, object_type, oid}
     -- system privileges of admin and guest can't be revoked
+
+    local origins
     if tuple == nil then
+        origins = {[DEFAULT_ORIGIN] = 0}
+    elseif tuple[PRIV_OPTS_FIELD_ID] == nil or
+           tuple[PRIV_OPTS_FIELD_ID].origins == nil then
+        origins = {[DEFAULT_ORIGIN] = tuple.privilege}
+    else
+        origins = tuple[PRIV_OPTS_FIELD_ID].origins
+    end
```

Comment:
We have such a code as in `grant` as well as in `revoke`. Can we extract it into a function like `origins_from_tuple` or `privilege_origins_from_tuple`?

I would also leave a comment that the function returns the origins in a full form (as a contrary to the normalized form stored in the _priv space) to ease further processing.

Reply:
Done.

Reply:
Thanks!